File: web/src/stores/config-store.ts
import { create } from 'zustand';
import type { AppConfig } from '@/types/config';
import {
  fetchConfig,
  saveConfig as apiSaveConfig,
  applyConfig as apiApplyConfig,
} from '@/lib/api';

function deepEqual(a: unknown, b: unknown): boolean {
  return JSON.stringify(a) === JSON.stringify(b);
}

function deepClone<T>(obj: T): T {
  return JSON.parse(JSON.stringify(obj));
}

interface ConfigState {
  /** 服务端已保存的配置（基线） */
  config: AppConfig | null;
  /** 用户编辑中的草稿 */
  draft: AppConfig | null;
  /** 正在加载配置 */
  loading: boolean;
  /** 正在保存 */
  saving: boolean;
  /** 正在应用 */
  applying: boolean;
  /** 加载错误 */
  error: string | null;
}

interface ConfigActions {
  loadConfig: () => Promise<void>;
  updateDraft: (updater: (prev: AppConfig) => AppConfig) => void;
  setDraft: (config: AppConfig) => void;
  save: (overrideDraft?: AppConfig) => Promise<boolean>;
  apply: () => Promise<boolean>;
  saveAndApply: (overrideDraft?: AppConfig) => Promise<boolean>;
  reset: () => void;
}

export type ConfigStore = ConfigState & ConfigActions;

export const useConfigStore = create<ConfigStore>((set, get) => ({
  config: null,
  draft: null,
  loading: true,
  saving: false,
  applying: false,
  error: null,

  loadConfig: async () => {
    set({ loading: true, error: null });
    try {
      const cfg = await fetchConfig();
      set({ config: cfg, draft: deepClone(cfg), loading: false });
    } catch (err) {
      set({
        error: err instanceof Error ? err.message : '加载配置失败',
        loading: false,
      });
    }
  },

  updateDraft: (updater) => {
    const { draft } = get();
    if (draft) {
      set({ draft: updater(deepClone(draft)) });
    }
  },

  setDraft: (config) => {
    set({ draft: deepClone(config) });
  },

  save: async (overrideDraft?) => {
    const draftToSave = overrideDraft ?? get().draft;
    if (!draftToSave) return false;
    set({ saving: true });
    try {
      await apiSaveConfig(draftToSave);
      const next = deepClone(draftToSave);
      set({ config: next, draft: deepClone(next), saving: false });
      return true;
    } catch (err) {
      set({ saving: false });
      throw err;
    }
  },

  apply: async () => {
    set({ applying: true });
    try {
      await apiApplyConfig();
      set({ applying: false });
      return true;
    } catch (err) {
      set({ applying: false });
      throw err;
    }
  },

  saveAndApply: async (overrideDraft?) => {
    const saved = await get().save(overrideDraft);
    if (!saved) return false;
    return get().apply();
  },

  reset: () => {
    const { config } = get();
    if (config) set({ draft: deepClone(config) });
  },
}));

/** 选择器：草稿是否与基线不同 */
export const selectIsDirty = (s: ConfigStore) =>
  s.config !== null && s.draft !== null && !deepEqual(s.config, s.draft);

File: macos/LocalRouter/Stores/ConfigStore.swift
import Foundation

@Observable
final class ConfigStore {
    var config: AppConfig?
    var draft: AppConfig?
    var loading = true
    var saving = false
    var applying = false
    var error: String?
    var showDiffSheet = false
    var showRawEditor = false
    var diffMode: DiffMode = .view

    var showResultAlert = false
    var resultAlertSuccess = true
    var resultAlertMessage = ""

    // 重启确认（监听地址变更后需要重启服务才能生效）
    var showRestartAlert = false
    var restartListenHost = ""
    var restartListenPort = 0

    enum DiffMode {
        case view
        case save
        case saveAndApply
    }

    var isDirty: Bool {
        guard let config, let draft else { return false }
        return config != draft
    }

    func loadConfig(api: APIClient) async {
        loading = true
        error = nil
        do {
            let loaded = try await api.fetchConfig()
            config = loaded
            draft = loaded
        } catch {
            self.error = error.localizedDescription
        }
        loading = false
    }

    func updateDraft(_ updater: (inout AppConfig) -> Void) {
        guard var d = draft else { return }
        updater(&d)
        draft = d
    }

    func save(api: APIClient, overrideDraft: AppConfig? = nil) async throws {
        let configToSave = overrideDraft ?? draft
        guard let configToSave else { return }
        saving = true
        defer { saving = false }
        try await api.saveConfig(configToSave)
        config = configToSave
        draft = configToSave
    }

    @discardableResult
    func apply(api: APIClient) async throws -> ApplyResult {
        applying = true
        defer { applying = false }
        return try await api.applyConfig()
    }

    @discardableResult
    func saveAndApply(api: APIClient, overrideDraft: AppConfig? = nil) async throws -> ApplyResult {
        try await save(api: api, overrideDraft: overrideDraft)
        return try await apply(api: api)
    }

    /// apply 后若监听地址变更，处理重启提示/结果展示。
    func handleApplyResult(_ result: ApplyResult, successMessage: String) {
        guard result.restartRequired else {
            showResult(success: true, message: successMessage)
            return
        }
        if result.canRestart {
            restartListenHost = result.listenHost ?? ""
            restartListenPort = result.listenPort ?? 0
            showRestartAlert = true
        } else {
            showResult(
                success: true,
                message: "配置已保存并应用，但监听地址变更需手动重启服务（local-router restart）后生效"
            )
        }
    }

    func showResult(success: Bool, message: String) {
        resultAlertSuccess = success
        resultAlertMessage = message
        showResultAlert = true
    }

    func reset() {
        draft = config
    }

    func resetAll() {
        config = nil
        draft = nil
        loading = true
        error = nil
    }
}
